### ISSUE TITLE
Getting "Invalid Header provided" when accessing Github API

### ISSUE THREAD

**Issue by airportyh** [hyperium/hyper#594]
I am getting this error when trying to execute this program:

``` rust
extern crate hyper;

use std::io::Read;
use hyper::Client;
use hyper::header::Connection;

fn main() {
  let client = Client::new();
  let maybe_res = client
    .get("https://api.github.com/users/seanmonstar")
    .header(Connection::close())
    .send();
  match maybe_res {
    Ok(mut res) => {
      let mut body = String::new();
      let _  = res.read_to_string(&mut body);
      println!("{}", body);
    },
    Err(e) => println!("{}", e)
  }
}
```

The output is:
```
$ cargo run
   Compiling gh-pages v0.1.0 (file:///Users/airportyh/Home/Code/gh-pages)
     Running `target/debug/gh-pages`
Invalid Header provided
```

It works with other pages, just not the Github API. My first instinct is it might have something to do with HTTPS, but I don't want to jump to any conclusions.

**Comment by mlalic** [comment]
This seems to be a resurgence of #361...

Basically, Github does not delimit their response header lines with CRLF (only \n) and hyper can't handle only the linefeed.

Interestingly enough, they *do* properly use the CRLF for success responses; it is only when they're returning the 403 that the response uses only the LF. In this case, you're getting a 403 because Github expects a User-Agent header to be included with every request, otherwise it's "Forbidden".

---

Basically the following test fails:

```rust
    #[test]
    fn test_no_crlf_only_lf() {
        let mut raw = MockStream::with_input(b"HTTP/1.0 403 Forbidden\nCache-Control: no-cache\nConnection: close\nContent-Type: text/html\n\nRequest forbidden by administrative rules. Please make sure your request has a User-Agent header (http://developer.github.com/v3/#user-agent-required). Check https://developer.github.com for other possible causes.\n");
        let mut buf = BufReader::new(&mut raw);
        let res = parse_response(&mut buf);

        assert!(res.is_ok());
        assert_eq!(res.unwrap().subject.1, "Forbidden");
    }
```

and this one passes:

```rust
    #[test]
    fn test_crlf() {
        let mut raw = MockStream::with_input(b"HTTP/1.0 403 Forbidden\r\nCache-Control: no-cache\r\nConnection: close\r\nContent-Type: text/html\r\n\r\nRequest forbidden by administrative rules. Please make sure your request has a User-Agent header (http://developer.github.com/v3/#user-agent-required). Check https://developer.github.com for other possible causes.\n");
        let mut buf = BufReader::new(&mut raw);
        let res = parse_response(&mut buf);

        assert!(res.is_ok());
        assert_eq!(res.unwrap().subject.1, "Forbidden");
    }
```

The raw bytes given to the stream in the former correspond exactly to the raw response that gh returns when no UA is included in the request.

**Comment by airportyh** [comment]
Cool! Thanks for the detailed workaround. I will give it try.

**Comment by airportyh** [comment]
The workaround worked. Thank you!

**Comment by seanmonstar** [comment]
Aha, it was a specific bug in httparse, when the status-reason was followed `\n`. It would incorrectly consume 1 too many bytes, and that threw off all the rest of the parsing. Fixed in httparse 0.1.5, so all that is required is `cargo update`.
